### PR TITLE
Feat/stop loss

### PR DESCRIPTION
**Pull request by xmatthias**
This PR is based on the work from @berlinguyinca in #651, which was closed unintentionally after deleting the merged ccxt branch.
Unfortunately cherry-picking was close to impossible as the code changed a ton since.

## Summary
Add support for trailing stop loss
Solve the issue: #90, replace #216

## Quick changelog

- database based tracking of dynamic stop loss
- able to enable or disable this feature
- improve migrate script to be more flexible for new columns

### CODE REVIEW

File: freqtrade/persistence.py [hunk 1]
```diff
@@ -66,25 +66,45 @@ def has_column(columns, searchname: str) -> bool:
...
     """
     Checks if migration is necessary and migrates if necessary
     """
     inspector = inspect(engine)
 
     cols = inspector.get_columns('trades')
+    tabs = inspector.get_table_names()
+    table_back_name = 'trades_bak'
+    i = 0
+    while table_back_name in tabs:
```
Comment (gcarq) [inline freqtrade/persistence.py]: We could use `enumerate` here, which seems more pythonic, e.g.:
```
for i, table_back_name in enumerate(tabs):
    ...
```

File: freqtrade/persistence.py [hunk 2]
```diff
@@ -161,6 +184,51 @@ def __repr__(self):
             arrow.get(self.open_date).humanize() if self.is_open else 'closed'
         )
 
+    def adjust_stop_loss(self, current_price: float, stoploss: float, initial: bool =False):
+        """
+        this adjusts the stop loss to it's most recently observed setting
+        :param current_price:
```
Comment (gcarq) [inline freqtrade/persistence.py]: Please remove or write some text for `:param and :return` docs

File: freqtrade/persistence.py [hunk 3]
```diff
@@ -161,6 +184,51 @@ def __repr__(self):
...
+
+        # keeping track of the highest observed rate for this trade
+        if self.max_rate is None:
+            self.max_rate = current_price
+        else:
+            if current_price > self.max_rate:
+                self.max_rate = current_price
+
+        # no stop loss assigned yet
+        if self.stop_loss is None or self.stop_loss == 0:
```
Comment (gcarq) [inline freqtrade/persistence.py]: Can be simplified to `if not self.stop_loss`:
```
>>> stop_loss = None
>>> bool(stop_loss)
False
>>> stop_loss = 0
>>> bool(stop_loss)
False
```

File: freqtrade/constants.py [hunk 4]
```diff
@@ -61,6 +61,8 @@
             'minProperties': 1
         },
         'stoploss': {'type': 'number', 'maximum': 0, 'exclusiveMaximum': True},
+        'trailing_stop': {'type': 'boolean'},
+        'trailing_stop_positive': {'type': 'number', 'minimum': 0},
```
Comment (gcarq) [inline freqtrade/constants.py]: I think we should also define an upper limit of `1` for this property

**Comment by gcarq** [comment]
@xmatthias I'm not sure why the CI flake8 succeeded, because I get a warning on my local machine:
```
$ flake8 freqtrade
freqtrade/persistence.py:187:86: E252 missing whitespace around parameter equals

$ flake8 --version
3.5.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.6.5 on Linux
```

File: freqtrade/persistence.py (hunk 1 above)
Comment (xmatthias) [reply freqtrade/persistence.py]: beeing honest, i didnt expect this to work when `tabs` is updated within the for loop - but it does indeed work.

**Comment by xmatthias** [comment]
i wonder myself why flake does not warn on this - but i don't get any warning myself (on the same flake version, and on linux too).

File: freqtrade/persistence.py [hunk 5]
```diff
@@ -80,7 +80,7 @@ def check_migrate(engine) -> None:
     tabs = inspector.get_table_names()
     table_back_name = 'trades_bak'
     i = 0
```
Comment (gcarq) [inline freqtrade/persistence.py]: `i = 0` and `i += 1` can be removed now
